packages/headless/src/features/folding/folding-slice.ts

```diff
@@ -0,0 +1,125 @@
+import {createReducer} from '@reduxjs/toolkit';
+import {Result} from '../../api/search/search/result';
+import {isArray, removeDuplicates} from '../../utils/utils';
+import {executeSearch, fetchMoreResults} from '../search/search-actions';
+import {registerFolding} from './folding-actions';
+import {
+  FoldedResult,
+  FoldingFields,
+  getFoldingInitialState,
+} from './folding-state';
+
+interface ResultWithFolding extends Result {
+  parentResult?: ResultWithFolding;
+  childResults?: ResultWithFolding[];
+}
+
+function getRawValue(result: ResultWithFolding, field: string) {
+  const rawValue = result.raw[field] as string | [string] | undefined;
+  if (isArray(rawValue)) {
+    return rawValue[0];
+  }
+  return rawValue;
+}
+
+function areDefinedAndEqual<T>(
+  value1: T | undefined,
+  value2: T | undefined
+): boolean {
+  return (value1 || value2) !== undefined && value1 === value2;
+}
+
+function flattenResultTree(result: ResultWithFolding): ResultWithFolding[] {
+  const results = [result];
+  if (result.parentResult) {
+    results.unshift(result.parentResult);
+  }
+  if (result.childResults?.length) {
+    results.push(
+      ...result.childResults.flatMap((childResult) =>
+        flattenResultTree(childResult)
+      )
+    );
+  }
+  return results;
+}
+
+function foldResult(
+  source: ResultWithFolding,
+  results: ResultWithFolding[],
+  fields: FoldingFields
+): FoldedResult {
+  const sourceChildValue = getRawValue(source, fields.child);
+  return {
+    ...source,
+    children: sourceChildValue
+      ? results
+          .filter(
+            (result) =>
+              getRawValue(result, fields.child) !==
+                getRawValue(source, fields.child) &&
+              getRawValue(result, fields.parent) === sourceChildValue
+          )
+          .map((result) => foldResult(result, results, fields))
+      : [],
+  };
+}
+
+function createCollection(
+  relevantResult: ResultWithFolding,
```

Comment:
Would `rootResult` be a better name here? Or is there a special significance to the word `relevant` in the context of folding?

Reply:
There is some significance to the word `relevant` since the reason this result is the root is because it's the most relevant. I'm hesitant to use `rootResult` since it's only the root of the results returned by the API, once folded it's likely not to be the root anymore.

Reply:
I see, `root` would be misleading, I agree. Maybe some other ideas: `unprocessedResult`, `rawResult`